## Teoria da Probabilidade

### Trabalho sobre Simulação: 
#### (8/6/2019)


A descrição abaixo assume o uso de uma planilha para fazer o trabalho e apresentar seus resultados.  Se você optar por outro ambiente, além da planilha de resultados, entregue um relatório descrevendo cada etapa da tarefa, incluindo o código empregado.


Obtenha, na página do curso, o arquivo ```simulacao-base-2019.xls``` contendo a planilha de simulação (fila) e a planilha dados, que contém os valores de **f**, $m$, $a$, $M$ e $N$ com os quais você deve trabalhar, além do seu número id de identificação.  Obtenha também a planilha ```resultados-2019.xls``` , na qual você deve apresentar os resultados. Transcreva os dados para os locais indicados nas planilhas (as células a serem preenchidas estão em verde ou amarelo).  

Devolva, para  paulo.carvalho@fgv.br, somente a planilha de resultados renomeada como ```id.xls```, onde id é o seu número de identificação (escreva também o seu nome e seu número de identificação na planilha).

---

   #### 1.	Uma empresa tem um único atendente para receber seus clientes, que chegam para serem atendidos segundo um Processo de Poisson com média de uma chegada a cada $10$ minutos. Há dois tipos de cliente, os que desejam somente retirar um produto e os que necessitam um atendimento mais demorado:
   
  #### • Os do primeiro tipo, que representam uma fração __f__ do total de clientes, têm um tempo de atendimento entre 2 e 4 minutos, dado por $2 + 2X$, onde $X$ é uma variável aleatória com densidade $$f_X(x)   = (1+a) x^a , \forall x \in (0,1).$$ 
  #### • Os demais têm tempo de atendimento com distribuição exponencial com média $m$ minutos.

    a)	Programe a sua planilha para produzir os tempos de chegada e os tempos de atendimento para 10.000 clientes e, a partir daí, os demais dados (início do atendimento, fim do atendimento, etc.) para cada cliente. Calcule também as médias e desvios-padrão para as colunas indicadas.

    b)	Se você fez sua simulação usando a planilha, copie as primeiras 8 linhas da planilha de simulação para a planilha de resultados (copie todo o conteúdo das células, não somente os valores)

    c)	Calcule (analiticamente) o tempo médio de atendimento para os clientes e compare este valor com a média dos tempos de atendimento obtida na planilha. Transcreva estes valores para o local indicado na planilha (ao copiar os valores da sua planilha, cole apenas os valores).

    d)	De acordo com o seu experimento, em média quanto um cliente espera para começar a ser atendido?  Qual é o desvio-padrão desse tempo de espera? Que fração dos clientes tem que esperar para ser atendido? Que fração tem que esperar mais de 3 minutos?   Transcreva estes valores para o local indicado na planilha, copiando apenas os valores.
    
    e)	Qual é a fração do tempo em que um atendente fica ocioso?
  

---

#### 2.    Quantas figurinhas devem ser compradas até completar um álbum com $N$ figurinhas, supondo que não haja trocas de figurinhas? Simule $100$ repetições deste processo, supondo que as figurinhas sejam compradas individualmente e que não haja figurinhas difíceis.

    a)   Calcule a média e o desvio padrão obtido nas 100 realizações e transcreva-os para a planilha de resultados (somente os valores).  Obtenha também o valor exato da média e do desvio-padrão e compare com o resultado da simulação. 
    
    b)   Usando a simulação, estime a probabilidade de que o número de figurinhas compradas seja maior que M.  

___

 Importando os valores da planilha `simulacao_base_2019.xlsx`:

In [1]:
import numpy as np
import pandas as pd

simulacao = 'simulacao_base_2019.xlsx'

In [2]:
valores = pd.read_excel(simulacao)

valores.columns = valores.values[1]

valores = valores.drop(range(45), axis = 0)

valores = valores.drop(range(46,81), axis = 0)

Rafael = valores

Rafael

Nome  ID    f     a    m    N     M
45  Rafael Martins Kovashikawa  44  0.5 -0.55  5.5  635  4500

***
---

## Questão 1:

***

Para o primeiro tipo de cliente para um fração **f** = 50%:

Temos que $a = -0.55$, então: $$ f_X(x) = (0.45)x^{-0.55} ,\ \forall x \in (0,1) ,$$

Então, $$Y = 2 + 2X.$$

Logo, temos que: $f_Y(y)= \frac{f_X(y)}{g'(y)} \iff f_X(y)=(.45)\big(1+ \frac {y}{2}\big)^{-.55} ;\ g'(y)= 2$ 

$$\Longrightarrow f_Y(y)= (0.225)\Bigg(1+ \frac {y}{2}\Bigg),\ \forall y \in (2,4)$$

***

Para o segundo tipo de cliente:

Temos que $\lambda = 5.5$, tal que $\mu = \frac {1}{\lambda} \Longrightarrow \mu = \frac{1}{5.5} $,

então: 

$$ f_Z(z) = \frac{e^{ \frac{\Large-z\Large}{5.5}}}{5.5},\ \forall z \in [0,\infty )$$

***

### Letra a)

    Programe a sua planilha para produzir os tempos de chegada e os tempos de atendimento para 10.000 clientes e, a partir daí, os demais dados (início do atendimento, fim do atendimento, etc.) para cada cliente. Calcule também as médias e desvios-padrão para as colunas indicadas.
    
 __Solução:__
    
Criar um pd.DataFrame com os campos : 
    
    Cliente, Aleatorio 1, Intervalo de chegada, Instante de Chegada, Inicio do Atendimento, Aleatorio 2, Tempo de Serviço Rápido, Tempo de Serviço Longo, Aleatorio 3, Cliente Rápido?, Tempo de Serviço, Fim do Atendimento, Tempo de Espera, Tempo Ocioso, Tem que esperar?, Espera > 3 min?
    
Aonde clientes = range(1,10000), médias ($\hat \lambda_i$) e desvios padrão ($\hat \sigma$) para as colunas:
    
    Tempo de Serviço, Fim do Atendimento, Tempo de Espera, Tempo Ocioso, Tem que esperar?, Espera > 3 min?

In [3]:
# criando a tabela (pandas.DataFrame)

simulacao = pd.DataFrame(columns=['Cliente', 'Aleatório 1', 'Intervalo de chegada',
                                 'Instante de Chegada', 'Início do Atendimento', 
                                 'Aleatório 2', 'Tempo de Serviço Rápido', 
                                 'Tempo de Serviço Longo', 'Aleatório 3', 
                                 'Cliente Rápido?', 'Tempo de Serviço', 
                                 'Fim do Atendimento', 'Tempo de Espera', 
                                 'Tempo Ocioso', 'Tem que esperar?', 'Espera > 3 min?']
                        )

simulacao['Cliente'] = range(1,10001)


simulacao.tail()

Cliente Aleatório 1 Intervalo de chegada Instante de Chegada  \
9995     9996         NaN                  NaN                 NaN   
9996     9997         NaN                  NaN                 NaN   
9997     9998         NaN                  NaN                 NaN   
9998     9999         NaN                  NaN                 NaN   
9999    10000         NaN                  NaN                 NaN   

     Início do Atendimento Aleatório 2 Tempo de Serviço Rápido  \
9995                   NaN         NaN                     NaN   
9996                   NaN         NaN                     NaN   
9997                   NaN         NaN                     NaN   
9998                   NaN         NaN                     NaN   
9999                   NaN         NaN                     NaN   

     Tempo de Serviço Longo Aleatório 3 Cliente Rápido? Tempo de Serviço  \
9995                    NaN         NaN             NaN              NaN   
9996                    NaN         NaN             NaN              NaN   
9997                    NaN         NaN             NaN              NaN   
9998                    NaN         NaN             NaN              NaN   
9999                    NaN         NaN             NaN              NaN   

     Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
9995                NaN             NaN          NaN              NaN   
9996                NaN             NaN          NaN              NaN   
9997                NaN             NaN          NaN              NaN   
9998                NaN             NaN          NaN              NaN   
9999                NaN             NaN          NaN              NaN   

     Espera > 3 min?  
9995             NaN  
9996             NaN  
9997             NaN  
9998             NaN  
9999             NaN

como __f__ $= 0.5$, teremos uma $f.d.p$ de $W \sim Bernoulli(0.5)$ que distribuirá a quantidade de clientes _demorados (com $Y \sim Poisson(10)$) ou rápidos (com $Z \sim Expo(\frac {1}{5.5})$)_

Outrossim, $$ W \sim Binomial(1, 0.5) $$

Imputando os tipos de cliente na tabela:

In [4]:
# usando a fdp binomial da biblioteca Pandas como pd
# e atribuindo-a para definir o tipo de cliente atendido:

tipo = []

for cliente in simulacao['Cliente']:

    cliente = []
    cliente = np.random.binomial(size=1, n=1, p= 0.5)
    tipo.append(cliente[0])
    
    
simulacao['Cliente Rápido?'] = tipo
simulacao.tail()

Cliente Aleatório 1 Intervalo de chegada Instante de Chegada  \
9995     9996         NaN                  NaN                 NaN   
9996     9997         NaN                  NaN                 NaN   
9997     9998         NaN                  NaN                 NaN   
9998     9999         NaN                  NaN                 NaN   
9999    10000         NaN                  NaN                 NaN   

     Início do Atendimento Aleatório 2 Tempo de Serviço Rápido  \
9995                   NaN         NaN                     NaN   
9996                   NaN         NaN                     NaN   
9997                   NaN         NaN                     NaN   
9998                   NaN         NaN                     NaN   
9999                   NaN         NaN                     NaN   

     Tempo de Serviço Longo Aleatório 3  Cliente Rápido? Tempo de Serviço  \
9995                    NaN         NaN                0              NaN   
9996                    NaN         NaN                0              NaN   
9997                    NaN         NaN                1              NaN   
9998                    NaN         NaN                0              NaN   
9999                    NaN         NaN                1              NaN   

     Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
9995                NaN             NaN          NaN              NaN   
9996                NaN             NaN          NaN              NaN   
9997                NaN             NaN          NaN              NaN   
9998                NaN             NaN          NaN              NaN   
9999                NaN             NaN          NaN              NaN   

     Espera > 3 min?  
9995             NaN  
9996             NaN  
9997             NaN  
9998             NaN  
9999             NaN

In [5]:
# criando os valores aleatórios

aleatorio = ['Aleatório 1', 'Aleatório 2', 'Aleatório 3']

expo = []

for cliente in simulacao['Cliente']:
    
    alea = []
    alea = np.random.rand(1,3)
    expo.append(alea[0])
    
simulacao[aleatorio] = expo
simulacao.tail()

Cliente  Aleatório 1 Intervalo de chegada Instante de Chegada  \
9995     9996     0.406024                  NaN                 NaN   
9996     9997     0.327835                  NaN                 NaN   
9997     9998     0.669384                  NaN                 NaN   
9998     9999     0.930643                  NaN                 NaN   
9999    10000     0.818776                  NaN                 NaN   

     Início do Atendimento  Aleatório 2 Tempo de Serviço Rápido  \
9995                   NaN     0.495248                     NaN   
9996                   NaN     0.284652                     NaN   
9997                   NaN     0.587488                     NaN   
9998                   NaN     0.555878                     NaN   
9999                   NaN     0.233838                     NaN   

     Tempo de Serviço Longo  Aleatório 3  Cliente Rápido? Tempo de Serviço  \
9995                    NaN     0.577855                0              NaN   
9996                    NaN     0.303510                0              NaN   
9997                    NaN     0.114276                1              NaN   
9998                    NaN     0.132335                0              NaN   
9999                    NaN     0.482321                1              NaN   

     Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
9995                NaN             NaN          NaN              NaN   
9996                NaN             NaN          NaN              NaN   
9997                NaN             NaN          NaN              NaN   
9998                NaN             NaN          NaN              NaN   
9999                NaN             NaN          NaN              NaN   

     Espera > 3 min?  
9995             NaN  
9996             NaN  
9997             NaN  
9998             NaN  
9999             NaN

In [6]:
# criando os Intervalo de chegada & Instante de Chegada

intervalo = -10*np.log(1-simulacao['Aleatório 1'])

simulacao['Intervalo de chegada'] = intervalo

instante = [intervalo[0]]

for i in range(1,10001):
    
    h = i - 1
    inst = intervalo[i] + instante[h]
    instante.append(inst)

KeyError: 10000

In [7]:
# criando os Intervalo de chegada & Instante de Chegada

simulacao['Instante de Chegada'] = instante

atendimento = [instante[0]]

for i in range(1,10000):

    h = i - 1
    atend = instante[i] + atendimento[h]
    atendimento.append(atend)
    
simulacao['Inicio do Atendimento'] = atendimento

In [8]:
simulacao.head()

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0        1     0.427549              5.578287             5.578287   
1        2     0.962996             32.967324            38.545611   
2        3     0.541149              7.790293            46.335904   
3        4     0.509838              7.130188            53.466093   
4        5     0.067388              0.697665            54.163757   

  Início do Atendimento  Aleatório 2 Tempo de Serviço Rápido  \
0                   NaN     0.568843                     NaN   
1                   NaN     0.303098                     NaN   
2                   NaN     0.976082                     NaN   
3                   NaN     0.077555                     NaN   
4                   NaN     0.182399                     NaN   

  Tempo de Serviço Longo  Aleatório 3  Cliente Rápido? Tempo de Serviço  \
0                    NaN     0.821637                0              NaN   
1                    NaN     0.652828                1              NaN   
2                    NaN     0.172237                0              NaN   
3                    NaN     0.838362                0              NaN   
4                    NaN     0.878672                1              NaN   

  Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
0                NaN             NaN          NaN              NaN   
1                NaN             NaN          NaN              NaN   
2                NaN             NaN          NaN              NaN   
3                NaN             NaN          NaN              NaN   
4                NaN             NaN          NaN              NaN   

  Espera > 3 min?  Inicio do Atendimento  
0             NaN               5.578287  
1             NaN              44.123898  
2             NaN              90.459802  
3             NaN             143.925895  
4             NaN             198.089652

In [9]:
# criando os Tempo de Serviço Rápido & Tempo de Serviço Longo

rapid = 2 + 2*(simulacao['Aleatório 2']**(1/(1-0.55)))

simulacao['Tempo de Serviço Rápido'] = rapid

slow = (-5.5)*np.log(1-simulacao['Aleatório 2'])

simulacao['Tempo de Serviço Longo'] = slow

print(' tempo de atendimento devagar: \n\n', slow.describe(),
      '\n\n\n tempo de atendimento rápido: \n\n', rapid.describe())

simulacao.head()

 tempo de atendimento devagar: 

 count    10000.000000
mean         5.524555
std          5.528586
min          0.000276
25%          1.571573
50%          3.806223
75%          7.708760
max         60.741903
Name: Aleatório 2, dtype: float64 


 tempo de atendimento rápido: 

 count    10000.000000
mean         2.623597
std          0.595420
min          2.000000
25%          2.090672
50%          2.427568
75%          3.067235
max          3.999929
Name: Aleatório 2, dtype: float64


Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0        1     0.427549              5.578287             5.578287   
1        2     0.962996             32.967324            38.545611   
2        3     0.541149              7.790293            46.335904   
3        4     0.509838              7.130188            53.466093   
4        5     0.067388              0.697665            54.163757   

  Início do Atendimento  Aleatório 2  Tempo de Serviço Rápido  \
0                   NaN     0.568843                 2.570911   
1                   NaN     0.303098                 2.140927   
2                   NaN     0.976082                 3.895248   
3                   NaN     0.077555                 2.006815   
4                   NaN     0.182399                 2.045588   

   Tempo de Serviço Longo  Aleatório 3  Cliente Rápido? Tempo de Serviço  \
0                4.627056     0.821637                0              NaN   
1                1.986109     0.652828                1              NaN   
2               20.532150     0.172237                0              NaN   
3                0.443999     0.838362                0              NaN   
4                1.107591     0.878672                1              NaN   

  Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
0                NaN             NaN          NaN              NaN   
1                NaN             NaN          NaN              NaN   
2                NaN             NaN          NaN              NaN   
3                NaN             NaN          NaN              NaN   
4                NaN             NaN          NaN              NaN   

  Espera > 3 min?  Inicio do Atendimento  
0             NaN               5.578287  
1             NaN              44.123898  
2             NaN              90.459802  
3             NaN             143.925895  
4             NaN             198.089652

In [10]:
# Tempo de Serviço

tmp_serv = []

for cliente in range(10000):
    
    if simulacao['Cliente Rápido?'][cliente] == 1:
        
        tmp_serv.append(rapid[cliente])
        
    else:
        
        tmp_serv.append(slow[cliente])

simulacao['Tempo de Serviço'] = tmp_serv

simulacao.tail()

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
9995     9996     0.406024              5.209169         99247.797795   
9996     9997     0.327835              3.972508         99251.770303   
9997     9998     0.669384             11.067986         99262.838289   
9998     9999     0.930643             26.684828         99289.523117   
9999    10000     0.818776             17.080195         99306.603311   

     Início do Atendimento  Aleatório 2  Tempo de Serviço Rápido  \
9995                   NaN     0.495248                 2.419622   
9996                   NaN     0.284652                 2.122573   
9997                   NaN     0.587488                 2.613330   
9998                   NaN     0.555878                 2.542397   
9999                   NaN     0.233838                 2.079181   

      Tempo de Serviço Longo  Aleatório 3  Cliente Rápido?  Tempo de Serviço  \
9995                3.760282     0.577855                0          3.760282   
9996                1.842427     0.303510                0          1.842427   
9997                4.870195     0.114276                1          2.613330   
9998                4.464105     0.132335                0          4.464105   
9999                1.464990     0.482321                1          2.079181   

     Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
9995                NaN             NaN          NaN              NaN   
9996                NaN             NaN          NaN              NaN   
9997                NaN             NaN          NaN              NaN   
9998                NaN             NaN          NaN              NaN   
9999                NaN             NaN          NaN              NaN   

     Espera > 3 min?  Inicio do Atendimento  
9995             NaN           4.938889e+08  
9996             NaN           4.939882e+08  
9997             NaN           4.940874e+08  
9998             NaN           4.941867e+08  
9999             NaN           4.942860e+08

In [11]:
# Início do Atendimento

inicio = [instante[0]]

for cliente in range(1,10000):
    
    i = inicio[cliente-1] + tmp_serv[cliente-1]
    
    if i <= instante[cliente]:
        
        inicio.append(instante[cliente])
    else:
        
        inicio.append(i)
    

simulacao['Início do Atendimento'] = inicio

simulacao.head()

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0        1     0.427549              5.578287             5.578287   
1        2     0.962996             32.967324            38.545611   
2        3     0.541149              7.790293            46.335904   
3        4     0.509838              7.130188            53.466093   
4        5     0.067388              0.697665            54.163757   

   Início do Atendimento  Aleatório 2  Tempo de Serviço Rápido  \
0               5.578287     0.568843                 2.570911   
1              38.545611     0.303098                 2.140927   
2              46.335904     0.976082                 3.895248   
3              66.868055     0.077555                 2.006815   
4              67.312053     0.182399                 2.045588   

   Tempo de Serviço Longo  Aleatório 3  Cliente Rápido?  Tempo de Serviço  \
0                4.627056     0.821637                0          4.627056   
1                1.986109     0.652828                1          2.140927   
2               20.532150     0.172237                0         20.532150   
3                0.443999     0.838362                0          0.443999   
4                1.107591     0.878672                1          2.045588   

  Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
0                NaN             NaN          NaN              NaN   
1                NaN             NaN          NaN              NaN   
2                NaN             NaN          NaN              NaN   
3                NaN             NaN          NaN              NaN   
4                NaN             NaN          NaN              NaN   

  Espera > 3 min?  Inicio do Atendimento  
0             NaN               5.578287  
1             NaN              44.123898  
2             NaN              90.459802  
3             NaN             143.925895  
4             NaN             198.089652

In [13]:
# Fim do Atendimento

fim = []

for cliente in range(10000):
    
    i = inicio[cliente] + tmp_serv[cliente]
    fim.append(i)

simulacao['Fim do Atendimento'] = fim

simulacao.head()

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0        1     0.427549              5.578287             5.578287   
1        2     0.962996             32.967324            38.545611   
2        3     0.541149              7.790293            46.335904   
3        4     0.509838              7.130188            53.466093   
4        5     0.067388              0.697665            54.163757   

   Início do Atendimento  Aleatório 2  Tempo de Serviço Rápido  \
0               5.578287     0.568843                 2.570911   
1              38.545611     0.303098                 2.140927   
2              46.335904     0.976082                 3.895248   
3              66.868055     0.077555                 2.006815   
4              67.312053     0.182399                 2.045588   

   Tempo de Serviço Longo  Aleatório 3  Cliente Rápido?  Tempo de Serviço  \
0                4.627056     0.821637                0          4.627056   
1                1.986109     0.652828                1          2.140927   
2               20.532150     0.172237                0         20.532150   
3                0.443999     0.838362                0          0.443999   
4                1.107591     0.878672                1          2.045588   

   Fim do Atendimento Tempo de Espera Tempo Ocioso Tem que esperar?  \
0           10.205343             NaN          NaN              NaN   
1           40.686538             NaN          NaN              NaN   
2           66.868055             NaN          NaN              NaN   
3           67.312053             NaN          NaN              NaN   
4           69.357642             NaN          NaN              NaN   

  Espera > 3 min?  Inicio do Atendimento  
0             NaN               5.578287  
1             NaN              44.123898  
2             NaN              90.459802  
3             NaN             143.925895  
4             NaN             198.089652

In [14]:
# Tempo de Espera

espera = []

for cliente in range(10000):
    
    if inicio[cliente] <= instante[cliente]:
        
        espera.append(0)
        
    else:
        
        espera.append(inicio[cliente] - instante[cliente])

simulacao['Tempo de Espera'] = espera

# Tempo Ocioso

ocioso = [inicio[0]]

for cliente in range(1,10000):
    if inicio[cliente] <= fim[cliente - 1]:
        ocioso.append(0)
    else:
        ocioso.append(inicio[cliente] - fim[cliente-1])

#simulacao['Tempo de Espera'] = espera

#simulacao.head()

simulacao['Tempo Ocioso'] = ocioso

simulacao.head()

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0        1     0.427549              5.578287             5.578287   
1        2     0.962996             32.967324            38.545611   
2        3     0.541149              7.790293            46.335904   
3        4     0.509838              7.130188            53.466093   
4        5     0.067388              0.697665            54.163757   

   Início do Atendimento  Aleatório 2  Tempo de Serviço Rápido  \
0               5.578287     0.568843                 2.570911   
1              38.545611     0.303098                 2.140927   
2              46.335904     0.976082                 3.895248   
3              66.868055     0.077555                 2.006815   
4              67.312053     0.182399                 2.045588   

   Tempo de Serviço Longo  Aleatório 3  Cliente Rápido?  Tempo de Serviço  \
0                4.627056     0.821637                0          4.627056   
1                1.986109     0.652828                1          2.140927   
2               20.532150     0.172237                0         20.532150   
3                0.443999     0.838362                0          0.443999   
4                1.107591     0.878672                1          2.045588   

   Fim do Atendimento  Tempo de Espera  Tempo Ocioso Tem que esperar?  \
0           10.205343         0.000000      5.578287              NaN   
1           40.686538         0.000000     28.340268              NaN   
2           66.868055         0.000000      5.649367              NaN   
3           67.312053        13.401962      0.000000              NaN   
4           69.357642        13.148296      0.000000              NaN   

  Espera > 3 min?  Inicio do Atendimento  
0             NaN               5.578287  
1             NaN              44.123898  
2             NaN              90.459802  
3             NaN             143.925895  
4             NaN             198.089652

In [26]:
# criando as variáveis para o lambda chapeu:

𝜆hat = ['Tempo de Serviço', 'Tempo de Espera', 
     'Tempo Ocioso', 'Tem que esperar?', 'Espera > 3 min?']

simulacao[𝜆hat].describe()

Tempo de Serviço  Tempo de Espera  Tempo Ocioso
count      10000.000000     10000.000000  10000.000000
mean           4.056071         2.912769      5.874798
std            4.144760         6.221992      9.233895
min            0.000276         0.000000      0.000000
25%            2.046504         0.000000      0.000000
50%            2.664997         0.000000      1.566242
75%            3.917097         2.806523      8.538675
max           42.929947        58.815342     86.495933

In [39]:
t_esperar = ['não']

esperar3 = ['não']

for cliente in range(1, 10000):
    
    if inicio[cliente] <= instante[cliente]:
    
        t_esperar.append('não')
        
    else:
        
        t_esperar.append('sim')
        
    if inicio[cliente] - instante[cliente] <= 3:
    
        esperar3.append('não')
        
    else:
        
        esperar3.append('sim')

simulacao['Tem que esperar?'] = t_esperar

simulacao['Espera > 3 min?'] = esperar3

simulacao

Cliente  Aleatório 1  Intervalo de chegada  Instante de Chegada  \
0           1     0.427549              5.578287             5.578287   
1           2     0.962996             32.967324            38.545611   
2           3     0.541149              7.790293            46.335904   
3           4     0.509838              7.130188            53.466093   
4           5     0.067388              0.697665            54.163757   
5           6     0.845344             18.665489            72.829246   
6           7     0.398537              5.083899            77.913146   
7           8     0.679412             11.375985            89.289131   
8           9     0.455616              6.081006            95.370137   
9          10     0.198171              2.208600            97.578737   
10         11     0.276218              3.232648           100.811385   
11         12     0.732443             13.184242           113.995627   
12         13     0.686683             11.605409           125.601036   
13         14     0.850405             18.998264           144.599299   
14         15     0.529088              7.530835           152.130134   
15         16     0.500097              6.933406           159.063541   
16         17     0.375701              4.711264           163.774804   
17         18     0.971770             35.673743           199.448547   
18         19     0.591233              8.946101           208.394648   
19         20     0.862402             19.834184           228.228832   
20         21     0.768151             14.616703           242.845535   
21         22     0.384071              4.846239           247.691774   
22         23     0.698713             11.996911           259.688685   
23         24     0.286559              3.376554           263.065239   
24         25     0.574196              8.537772           271.603011   
25         26     0.349645              4.302369           275.905379   
26         27     0.437727              5.757678           281.663057   
27         28     0.560790              8.227783           289.890841   
28         29     0.987745             44.018354           333.909195   
29         30     0.507013              7.072727           340.981921   
...       ...          ...                   ...                  ...   
9970     9971     0.627132              9.865299         98974.216393   
9971     9972     0.388452              4.917613         98979.134006   
9972     9973     0.917512             24.951077         99004.085083   
9973     9974     0.182383              2.013609         99006.098692   
9974     9975     0.138102              1.486187         99007.584879   
9975     9976     0.522102              7.383574         99014.968453   
9976     9977     0.590167              8.920053         99023.888506   
9977     9978     0.384074              4.846281         99028.734787   
9978     9979     0.917848             24.991805         99053.726592   
9979     9980     0.872666             20.609431         99074.336024   
9980     9981     0.812233             16.725514         99091.061537   
9981     9982     0.754043             14.025987         99105.087525   
9982     9983     0.195093              2.170284         99107.257808   
9983     9984     0.063890              0.660226         99107.918034   
9984     9985     0.296938              3.523102         99111.441136   
9985     9986     0.106976              1.131417         99112.572554   
9986     9987     0.379581              4.773600         99117.346153   
9987     9988     0.027471              0.278556         99117.624709   
9988     9989     0.967227             34.181551         99151.806260   
9989     9990     0.856929             19.444161         99171.250421   
9990     9991     0.057276              0.589822         99171.840243   
9991     9992     0.365333              4.546547         99176.386789   
9992     9993     0.695757             11.899302         99188.2

***
***

### Letras c, d & e)

    c)	Calcule (analiticamente) o tempo médio de atendimento para os clientes e compare este valor com a média dos tempos de atendimento obtida na planilha. Transcreva estes valores para o local indicado na planilha (ao copiar os valores da sua planilha, cole apenas os valores).

    d)	De acordo com o seu experimento, em média quanto um cliente espera para começar a ser atendido?  Qual é o desvio-padrão desse tempo de espera? Que fração dos clientes tem que esperar para ser atendido? Que fração tem que esperar mais de 3 minutos?   Transcreva estes valores para o local indicado na planilha, copiando apenas os valores.
    
    e)	Qual é a fração do tempo em que um atendente fica ocioso?

In [36]:
# letra c)

print('tempo medio de atendimento aos cliente:\n\n E(A) = (0.5)EY + (0.5)EZ \n\n = (0.5)(2.620689655 + 5.5) \n\n = 4.060344828')

simulacao['Tempo de Serviço'].describe()

tempo medio de atendimento aos cliente:

 E(A) = (0.5)EY + (0.5)EZ 

 = (0.5)(2.620689655 + 5.5) 

 = 4.060344828


count    10000.000000
mean         4.056071
std          4.144760
min          0.000276
25%          2.046504
50%          2.664997
75%          3.917097
max         42.929947
Name: Tempo de Serviço, dtype: float64

In [18]:
# letra d)

simulacao['Tempo de Espera'].describe()

count    10000.000000
mean         2.912769
std          6.221992
min          0.000000
25%          0.000000
50%          0.000000
75%          2.806523
max         58.815342
Name: Tempo de Espera, dtype: float64

In [37]:
clientes_esperando = 0

for i in t_esperar:
    
    if i == 'sim':
        
        clientes_esperando = clientes_esperando + 1

print('fracao de clientes que esperam: ', clientes_esperando/10000)

clientes_esperando3 = 0

for i in esperar3:
    
    if i == 'sim':
        
        clientes_esperando3 = clientes_esperando3 + 1
        
print('fracao de clientes que esperam 3 minutos ou mais: ', clientes_esperando3/10000)

fracao de clientes que esperam:  0.4088
fracao de clientes que esperam 3 minutos ou mais:  0.2423


In [40]:
# letra e)

sum(ocioso)/fim[-1]

0.5915693912140958

***
***

In [41]:
Rafael

Nome  ID    f     a    m    N     M
45  Rafael Martins Kovashikawa  44  0.5 -0.55  5.5  635  4500

### Questão 2 

    Quantas figurinhas devem ser compradas até completar um álbum com N figurinhas, supondo que não haja trocas de figurinhas? Simule 100 repetições deste processo, supondo que as figurinhas sejam compradas individualmente e que não haja figurinhas difíceis.
    
    a)   Calcule a média e o desvio padrão obtido nas 100 realizações e transcreva-os para a planilha de resultados (somente os valores).  Obtenha também o valor exato da média e do desvio-padrão e compare com o resultado da simulação. 
    
    b)   Usando a simulação, estime a probabilidade de que o número de figurinhas compradas seja maior que M.
   
__Solucao:__

temos que $N = 635$ e $M = 4500$, ou seja, precisaremos de fazer uma tabela (*pandas.Dataframe*) com 